src/ng/compile.js

```diff
@@ -310,6 +316,39 @@
  *   they are waiting for their template to load asynchronously and their own compilation and linking has been
  *   suspended until that occurs.
  *
+ * ** $doCheck example **
+ *
+ * This example show how you might use `$doCheck` to customise the equality check of component inputs.
+ *
+ * <example name="doCheckExample" module="do-check-module">
+ *   <file name="index.html">
+ *     <div ng-init="items = []">
+ *       <button ng-click="items.push(items.length)">Add Item</button>
+ *       <button ng-click="items = []">Reset Items</button>
+ *       <pre>{{ items }}</pre>
+ *       <test items="items"></test>
+ *     </div>
+ *   </file>
+ *   <file name="app.js">
+ *      angular.module('do-check-module', [])
+ *        .component('test', {
+ *          bindings: { items: '<' },
+ *          template:
+ *            '<pre>{{ $ctrl.log | json }}</pre>',
+ *          controller: function() {
+ *            this.log = [];
+ *
+ *            this.$doCheck = function() {
+ *              if (this.items_ref !== this.items) { this.log.push('doCheck: items changed'); }
+ *              if (!angular.equals(this.items_clone, this.items)) { this.log.push('doCheck: items mutated'); }
+ *
+ *              this.items_clone = angular.copy(this.items);
+ *              this.items_ref = this.items;
+ *            };
```

Comment:
Another usecase I have in mind for `$doCheck()` is for implementing a poor man's immutability-based change detection:

```js
// In parent:
this.items = [];
this.items._version = 0;

this.addItem = function addItem(item) {
  this.items.push(item);
  this.items._version++;
};

// In child:
this.$doCheck = function $doCheck() {
  if (this.items._version !== this.oldVersion) {
    // `items` has been mutated
  }
  this.oldVersion = this.items._version;
};
```

(Although I am not sure this is a vaid usecase :grin:)